core/src/main/scala/kafka/cluster/Partition.scala

```diff
@@ -1571,14 +1620,26 @@ class Partition(val topicPartition: TopicPartition,
     error match {
       case Errors.OPERATION_NOT_ATTEMPTED =>
         // Since the operation was not attempted, it is safe to reset back to the committed state.
-        partitionState = CommittedPartitionState(proposedIsrState.isr, LeaderRecoveryState.RECOVERED)
+        partitionState = proposedIsrState.lastCommittedState
```

Comment:
When we receive OPERATION_NOT_ATTEMPTED or INELIGIBLE_REPLICA, the controller has not done anything with out request besides validating it. This is why, I think that it is safe to revert to the last committed state, assuming that the partition epoch has not changed in the meantime.

Is your question about those two error codes or a more general one?

Reply:
Right, the controller sending the error code has not executed anything. But what happens when you mix in retries and controller changes? It is possible that a previous/future controller did act on the request and we did not receive a response? I believe it works correctly for two reasons. First, we are validating the partition epoch on the controller before checking for ineligible replicas. So if a controller made the ISR change and then failed over before we got a response, our retry would result in INVALID_UPDATE_VERSION. Second, the raft layer guarantees the monotonicity of the controller epoch. In the same scenario where we have lost a response, this ensures that we could not retry the request against an older controller (which might still have the old partition epoch). Anyway, my point is that this is a little subtle and depends on external invariants.

Reply:
Thanks. I think that I understand what you mean but I have a hard to translate it into a good and succinct  comment that we could put here. If you have any suggestion, feel free. Otherwise, I will give it another try tomorrow.

Reply:
How about we collapse these two cases into one `case`. Then we can add a comment like this:

```scala
// Care must be taken when resetting to the last committed state since we may not
// know in general whether the request was applied or not taking into account retries
// and controller changes which might have occurred before we received the response.
// However, when the controller returns INELIGIBLE_REPLICA (or OPERATION_NOT_COMMITTED), 
// the controller is explicitly telling us 1) that the current partition epoch is correct, 
// and 2) that the request was not applied. Even if the controller that sent the response
// is stale, we are guaranteed from the monotonicity of the controller epoch that the 
// request could not have been applied by any past or future controller.
```

Reply:
Thanks, Jason!